### PR TITLE
[PDI-8931] SlaveServer Synchronized methods need to be examined

### PR DESCRIPTION
**Pull request by IvanNikolaychuk**
- Removed unnecessary synchronization, added lacked synchronization
- Test written
- Checkstyle applied

### CODE REVIEW

**Comment by IvanNikolaychuk** [comment]
@mchen-len-son , could you please review this PR?

**Comment by IvanNikolaychuk** [comment]
@mchen-len-son , looks like there are some problems with WN.
Works fine localy.

**Comment by mchen-len-son** [comment]
@mbatchelor can you please take a look at the PR?

**Comment by mbatchelor** [comment]
Not a fan actually. Instead of reducing synchronization, it was substantially increased which I believe will lead to a performance regression. I also will be making some inline comments - but I think this needs to be reworked using re-entrant read/write locks instead of sprinkling the synchronized keyword everywhere.

File: engine/src/org/pentaho/di/cluster/SlaveServer.java [hunk 1]
```diff
@@ -241,7 +242,7 @@ public LogChannelInterface getLogChannel() {
     return log;
   }
 
-  public String getXML() {
+  public synchronized String getXML() {
```
Comment (mbatchelor) [inline engine/src/org/pentaho/di/cluster/SlaveServer.java]: This makes absolutely no sense. This method creates a new StringBuilder, appends constants to the string, and returns the toString() of it. No actual instance variables referenced. This should absolutely not be synchronized!!!

File: engine/src/org/pentaho/di/cluster/SlaveServer.java [hunk 2]
```diff
@@ -266,13 +267,13 @@ public String getXML() {
     return xml.toString();
   }
 
-  public Object clone() {
+  public synchronized Object clone() {
```
Comment (mbatchelor) [inline engine/src/org/pentaho/di/cluster/SlaveServer.java]: This makes no sense either. The substantive "work" done here reading instance variables is done in replaceMeta. Yet, you synchronize here, and synchronize replaceMeta. This is not rational.

File: engine/src/org/pentaho/di/cluster/SlaveServer.java [hunk 3]
```diff
@@ -266,13 +267,13 @@ public String getXML() {
...
     SlaveServer slaveServer = new SlaveServer();
     slaveServer.replaceMeta( this );
     return slaveServer;
   }
 
-  public void replaceMeta( SlaveServer slaveServer ) {
+  public synchronized void replaceMeta( SlaveServer slaveServer ) {
```
Comment (mbatchelor) [inline engine/src/org/pentaho/di/cluster/SlaveServer.java]: a re-entrant write lock would make sense here.

File: engine/src/org/pentaho/di/cluster/SlaveServer.java [hunk 4]
```diff
@@ -443,14 +445,16 @@ public String getPortSpecification() {
     return portSpec;
   }
 
-  public String constructUrl( String serviceAndArguments ) throws UnsupportedEncodingException {
+  public synchronized String constructUrl( String serviceAndArguments ) throws UnsupportedEncodingException {
```
Comment (mbatchelor) [inline engine/src/org/pentaho/di/cluster/SlaveServer.java]: A readlock would make more sense than synchronizing the whole method.

File: engine/src/org/pentaho/di/cluster/SlaveServer.java [hunk 5]
```diff
@@ -604,25 +608,35 @@ public String sendExport( String filename, String type, String load ) throws Exc
...
-    String host = environmentSubstitute( this.hostname );
-    String phost = environmentSubstitute( this.proxyHostname );
-    String pport = environmentSubstitute( this.proxyPort );
-    String nonprox = environmentSubstitute( this.nonProxyHosts );
+    String hostName;
+    String proxyHost;
+    String proxyPort;
+    String nonProxyHosts;
+
+    synchronized ( this ) {
```
Comment (mbatchelor) [inline engine/src/org/pentaho/di/cluster/SlaveServer.java]: Again - a write lock would be much better here.

File: engine/src/org/pentaho/di/cluster/SlaveServer.java [hunk 6]
```diff
@@ -604,25 +608,35 @@ public String sendExport( String filename, String type, String load ) throws Exc
...
       }
-      client.getHostConfiguration().setProxy( phost, Integer.parseInt( pport ) );
+      client.getHostConfiguration().setProxy( proxyHost, Integer.parseInt( proxyPort ) );
     }
   }
 
   public void addCredentials( HttpClient client ) {
-    client.getState().setCredentials(
+    HttpState state = client.getState();
+    synchronized ( this ) {
```
Comment (mbatchelor) [inline engine/src/org/pentaho/di/cluster/SlaveServer.java]: A read lock makes more sense here.

File: engine/src/org/pentaho/di/cluster/SlaveServer.java [hunk 7]
```diff
@@ -1123,7 +1137,7 @@ public long getNextSlaveSequenceValue( String slaveSequenceName, long incrementV
     }
   }
 
-  public SlaveServer getClient() {
+  public synchronized SlaveServer getClient() {
```
Comment (mbatchelor) [inline engine/src/org/pentaho/di/cluster/SlaveServer.java]: a Read Lock (if necessary) on the variables makes more sense than synchronizing the entire method.

File: engine/src/org/pentaho/di/cluster/SlaveServer.java (hunk 1 above)
Comment (IvanNikolaychuk) [inline engine/src/org/pentaho/di/cluster/SlaveServer.java]: Sorry, Marc, you are wrong here.
Lets assume we have a a slaveServer shared among 2 threads (Thread-1 and Thread-2).
Lets assume getXml() method is not synchronized.

Thread-1 calls getXml() method and in the middle of XML-method execution Thread-2 calls replaceMeta method. replaceMeta begins mutating fields of sharedSlaveServer object.
As a result getXml() method can mix up it's variables (one half of variables can be from the set before replaceMeta call here, another half - after replaceMeta call).

Synchronization is needed here, agree?

Comment (pentaho-nbaker) [inline engine/src/org/pentaho/di/cluster/SlaveServer.java]: A ReadWrite Lock can accommodate this just fine. If a Read lock is in effect the write will wait. If a Write is in affect the read will wait. Concurrent reads are not mutexed

File: engine/src/org/pentaho/di/cluster/SlaveServer.java (hunk 2 above)
Comment (IvanNikolaychuk) [inline engine/src/org/pentaho/di/cluster/SlaveServer.java]: Same for here.
Lets assume clone is not synchronized here.
Lets assume that Thread-1 is currently executeing replaceMeta() and it changed half of the fields of sharedSlaveServer.

Thread-2 calls clone method and clone method passes this reference to replace method. But this is not consistent, becaue not all properties were replaced yet. So, we get half of the properties replaced, half - not yet. Thats why synchronization is needed either.

Agree?

**Comment by mbatchelor** [comment]
One last comment - if this class is really this volatile, why would you have synchronization around some things that look at variables and not all things looking at variables. For example, you synchronize on constructUrl (ostensibly because it reads hostname and other instance variables). However, you aren't synchronizing on getHostname() . . .  

So once again - the purpose of me writing this JIRA case back then was that I did a review of synchronization in PDI, and found places where it was inconsistent. This class was inconsistent before you started in it, and has become even more inconsistent and incoherent.

Closing this PR as it doesn't hit the mark.

File: engine/src/org/pentaho/di/cluster/SlaveServer.java (hunk 1 above)
Comment (IvanNikolaychuk) [inline engine/src/org/pentaho/di/cluster/SlaveServer.java]: Yeah, I agree that Reentrant lock would be a better solution.
I'm just saing that this this should be synchronized.

Comment (mbatchelor) [inline engine/src/org/pentaho/di/cluster/SlaveServer.java]: How would it make sense to have the rug pulled out from under Thread-1? Perhaps a ThreadLocal makes more sense...

Comment (mbatchelor) [inline engine/src/org/pentaho/di/cluster/SlaveServer.java]: Once again - this returns "      <slaveserver><cr> " - Nothing at all to synchronize here ... It's a constant.

Comment (IvanNikolaychuk) [inline engine/src/org/pentaho/di/cluster/SlaveServer.java]: I think read lock would be a good solution, as Nick said abouve.
But this deffinately needs synchronization

Comment (IvanNikolaychuk) [inline engine/src/org/pentaho/di/cluster/SlaveServer.java]: It references all the fields of slaveServer object

Comment (mbatchelor) [inline engine/src/org/pentaho/di/cluster/SlaveServer.java]: Then at a read lock - not synchronizing the whole method.

Comment (IvanNikolaychuk) [inline engine/src/org/pentaho/di/cluster/SlaveServer.java]: Agree.

File: engine/src/org/pentaho/di/cluster/SlaveServer.java (hunk 2 above)
Comment (IvanNikolaychuk) [inline engine/src/org/pentaho/di/cluster/SlaveServer.java]: What about this?

**Comment by IvanNikolaychuk** [comment]
About your last comment:

I synchronize on constructUrl , because it takes a bunch of parameters for it's work.
This parameters need to be taken from one sharedServer and should not be mutaded in the middle of operation from another thread.

Again, lets assume it's not synchronized. We call constructUrl method from Thread-1, and execute this line:
`String realHostname = environmentSubstitute( hostname );`

Thread-2 calles replace method on sharedSlaveServer. It mutates all fields.

Thread-1 continues it's work.
We reach line:

`if ( !StringUtils.isBlank( webAppName ) ) {
`
but webAppName is mutated already and is no longer refers to that hostname, that we added before Thread-2 called constructUrl.

**Pull request closed without merging**